src/vs/workbench/services/editor/common/customEditorLabelService.ts

```diff
@@ -0,0 +1,185 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Emitter, Event } from 'vs/base/common/event';
+import { ParsedPattern, parse as parseGlob } from 'vs/base/common/glob';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { isAbsolute, parse as parsePath, ParsedPath } from 'vs/base/common/path';
+import { relativePath } from 'vs/base/common/resources';
+import { URI } from 'vs/base/common/uri';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { IWorkspaceContextService, IWorkspaceFolder } from 'vs/platform/workspace/common/workspace';
+
+interface ICustomEditorLabelObject {
+	readonly [key: string]: string;
+}
+
+interface ICustomEditorLabelPattern {
+	readonly pattern: string;
+	readonly template: string;
+
+	readonly isAbsolutePath: boolean;
+	readonly parsedPattern: ParsedPattern;
+}
+
+export class CustomEditorLabelService extends Disposable implements ICustomEditorLabelService {
+
+	readonly _serviceBrand: undefined;
+
+	static readonly SETTING_ID_PATTERNS = 'workbench.editor.label.patterns';
+	static readonly SETTING_ID_ENABLED = 'workbench.editor.label.enabled';
```

Comment:
I find the setting name strange: my impression is that I enable or disable the editor label? But this is about custom labels with custom patterns.

Reply:
I understand that, maybe `workbench.editor.customLabels.enabled` is better

Reply:
Yeah.